README.md
# Ripple Plugin for Tailwind CSS

## Inspired By

An excellent [Codepen](https://codepen.io/finnhvman/pen/jLXKJw) by [Ben Szabo](https://codepen.io/finnhvman).

## Installation

```bash
npm install tailwindcss-ripple
```

## Usage

```js
// tailwind.config.js
{
  theme: {
    ripple: theme => ({
        colors: theme('colors')
    }),
  },
  plugins: [
    require('tailwindcss-ripple')()
  ],
}
```

The default configuration generates the following ripple effect for each color in your theme :

```css
...

.ripple-bg-gray-300 {
    background-color: #e2e8f0;
    background-position: center;
    transition: background 0.8s;
}

.ripple-bg-gray-300:hover {
    background: #a5b7d0 radial-gradient(circle, transparent 1%, #a5b7d0 1%) center/15000%;
}

.ripple-bg-gray-300:active {
    background-color: #e2e8f0;
    background-size: 100%;
    transition: background 0s;
}

...
```

Which you can then use in your HTML like this:

```html
<button class="ripple-bg-gray-300">
    Hover me for a lighter background, click me for a ripple effect
</button>
```

## The Ripple Effect

By default, the color generated for the ripple effect is a 20% darken of the supplied color. This can be customised by passing in a parameter in your tailwind config file.

```js
// tailwind.config.js
{
  theme: {
    ripple: theme => ({
        colors: theme('colors'),
        darken: 0.1
    }),
  },
  plugins: [
    require('tailwindcss-ripple')()
  ],
}
```

You can also change the transition properties on the created class and it's active state

```js
// tailwind.config.js
{
  theme: {
    ripple: theme => ({
        colors: theme('colors'),
        modifierTransition: 'background 0.2s',
        activeTransition: 'background 0.1s'
    }),
  },
  plugins: [
    require('tailwindcss-ripple')()
  ],
}
```

Would generate the following CSS

```css
...

.ripple-bg-gray-300 {
    background-color: #e2e8f0;
    background-position: center;
    transition: background 0.2s;
}

.ripple-bg-gray-300:hover {
    background: #a5b7d0 radial-gradient(circle, transparent 1%, #a5b7d0 1%) center/15000%;
}

.ripple-bg-gray-300:active {
    background-color: #e2e8f0;
    background-size: 100%;
    transition: background 0.1s;
}

...
```

[![Build Status](https://travis-ci.org/jamessessford/tailwindcss-ripple.svg?branch=master)](https://travis-ci.org/jamessessford/tailwindcss-ripple)
[![Coverage Status](https://coveralls.io/repos/github/jamessessford/tailwindcss-ripple/badge.svg?branch=master)](https://coveralls.io/github/jamessessford/tailwindcss-ripple?branch=master)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project mostly adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

-   Updated node dependencies

## [0.7.1] - 2021-05-09

-   Updated node dependencies

## [0.7.0] - 2021-05-09

-   Updated node dependencies

## [0.6.0] - 2021-02-06

-   Named colors can now be exported [Issue](https://github.com/jamessessford/tailwindcss-ripple/issues/5).

## [0.5.0] - 2020-06-13

-   Darken value has been added as parameter [Issue](https://github.com/jamessessford/tailwindcss-ripple/issues/3).

## [0.4.0] - 2020-04-07

-   Fixed build
-   Updated dev dependencies

## [0.2.0] - 2019-10-18

### CHANGED

-   Completed testing setup

## [0.1.0] - 2019-10-17

### ADDED

-   Testing and Coverage

## [0.0.1] - 2019-10-17

Test Release

[Unreleased]: https://github.com/jamessessford/tailwindcss-ripple/compare/0.7.1...HEAD
[0.7.1]: https://github.com/jamessessford/tailwindcss-ripple/compare/0.7.1...0.7.0
[0.7.0]: https://github.com/jamessessford/tailwindcss-ripple/compare/0.7.0...0.6.0
[0.6.0]: https://github.com/jamessessford/tailwindcss-ripple/compare/0.6.0...0.5.0
[0.5.0]: https://github.com/jamessessford/tailwindcss-ripple/compare/0.5.0...0.4.0
[0.4.0]: https://github.com/jamessessford/tailwindcss-ripple/compare/0.4.0...0.2.0
[0.2.0]: https://github.com/jamessessford/tailwindcss-ripple/compare/0.2.0...0.1.0
[0.1.0]: https://github.com/jamessessford/tailwindcss-ripple/compare/0.1.0...0.0.1
[0.0.1]: https://github.com/jamessessford/tailwindcss-ripple/releases/tag/0.0.1


test.js
const _ = require('lodash');
const cssMatcher = require('jest-matcher-css');
const postcss = require('postcss');
const tailwindcss = require('tailwindcss');
const ripplePlugin = require('./index.js');

const generatePluginCss = config => {
    return postcss(
        tailwindcss(
            _.merge(
                {
                    theme: {},
                    corePlugins: false,
                    plugins: [ripplePlugin()],
                },
                config
            )
        )
    )
        .process('@tailwind components', {
            from: undefined,
        })
        .then(result => {
            return result.css;
        });
};

const variablesCssSheet = `

:root {
    --test-color: dodgerblue;
}

@tailwind components;
`;


const generatePluginVariablesCss = config => {
    return postcss(
        tailwindcss(
            _.merge(
                {
                    theme: {},
                    corePlugins: false,
                    plugins: [ripplePlugin()],
                },
                config
            )
        )
    )
        .process(variablesCssSheet, {
            from: undefined,
        })
        .then(result => {
            return result.css;
        });
};

expect.extend({
    toMatchCss: cssMatcher,
});

test('The plugin will generate nothing if no colors are available', () => {
    return generatePluginCss().then(css => {
        expect(css).toMatchCss(``);
    });
});

test('The plugin works with named colors', () => {
    return generatePluginCss({
        theme: {
            ripple: {
                colors: {
                    dodgerblue: 'dodgerblue',
                },
            },
        },
    }).then(css => {
        expect(css).toMatchCss(`
                .ripple-bg-dodgerblue {
                    background-color: dodgerblue;
                    background-position: center;
                    transition: background 0.8s;
                }

                .ripple-bg-dodgerblue:hover {
                    background: #0074E4 radial-gradient(circle, transparent 1%, #0074E4 1%) center/15000%;
                }

                .ripple-bg-dodgerblue:active {
                    background-color: dodgerblue;
                    background-size: 100%;
                    transition: background 0s;
                }
            `);
    });
});

test('The plugin will generate a requested colour', () => {
    return generatePluginCss({
        theme: {
            ripple: {
                colors: {
                    black: '#000',
                },
            },
        },
    }).then(css => {
        expect(css).toMatchCss(`
                .ripple-bg-black {
                    background-color: #000;
                    background-position: center;
                    transition: background 0.8s;
                }

                .ripple-bg-black:hover {
                    background: #000000 radial-gradient(circle, transparent 1%, #000000 1%) center/15000%;
                }

                .ripple-bg-black:active {
                    background-color: #000;
                    background-size: 100%;
                    transition: background 0s;
                }
            `);
    });
});

test('The plugin will generate a different colour', () => {
    return generatePluginCss({
        theme: {
            ripple: {
                colors: {
                    'purple-300': '#d6bcfa',
                },
            },
        },
    }).then(css => {
        expect(css).toMatchCss(`
                .ripple-bg-purple-300 {
                    background-color: #d6bcfa;
                    background-position: center;
                    transition: background 0.8s;
                }

                .ripple-bg-purple-300:hover {
                    background: #A46AF4 radial-gradient(circle, transparent 1%, #A46AF4 1%) center/15000%;
                }

                .ripple-bg-purple-300:active {
                    background-color: #d6bcfa;
                    background-size: 100%;
                    transition: background 0s;
                }
            `);
    });
});

test('The plugin will generate colors from a map', () => {
    return generatePluginCss({
        theme: {
            ripple: {
                colors: {
                    map: {
                        one: '#000',
                        two: '#fff',
                    },
                },
            },
        },
    }).then(css => {
        expect(css).toMatchCss(`
                .ripple-bg-map-one {
                    background-color: #000;
                    background-position: center;
                    transition: background 0.8s;
                }

                .ripple-bg-map-one:hover {
                    background: #000000 radial-gradient(circle, transparent 1%, #000000 1%) center/15000%;
                }

                .ripple-bg-map-one:active {
                    background-color: #000;
                    background-size: 100%;
                    transition: background 0s;
                }

                .ripple-bg-map-two {
                    background-color: #fff;
                    background-position: center;
                    transition: background 0.8s;
                }

                .ripple-bg-map-two:hover {
                    background: #CCCCCC radial-gradient(circle, transparent 1%, #CCCCCC 1%) center/15000%;
                }

                .ripple-bg-map-two:active {
                    background-color: #fff;
                    background-size: 100%;
                    transition: background 0s;
                }
            `);
    });
});

test('The plugin can accept a parameter for the darken value', () => {
    return generatePluginCss({
        theme: {
            ripple: {
                colors: {
                    map: {
                        one: '#000',
                        two: '#fff',
                    },
                },
                darken: 0.1
            },
        },
    }).then(css => {
        expect(css).toMatchCss(`
                .ripple-bg-map-one {
                    background-color: #000;
                    background-position: center;
                    transition: background 0.8s;
                }

                .ripple-bg-map-one:hover {
                    background: #000000 radial-gradient(circle, transparent 1%, #000000 1%) center/15000%;
                }

                .ripple-bg-map-one:active {
                    background-color: #000;
                    background-size: 100%;
                    transition: background 0s;
                }

                .ripple-bg-map-two {
                    background-color: #fff;
                    background-position: center;
                    transition: background 0.8s;
                }

                .ripple-bg-map-two:hover {
                    background: #E6E6E6 radial-gradient(circle, transparent 1%, #E6E6E6 1%) center/15000%;
                }

                .ripple-bg-map-two:active {
                    background-color: #fff;
                    background-size: 100%;
                    transition: background 0s;
                }
            `);
    });
});

test('The plugin uses the darken value', () => {
    return generatePluginCss({
        theme: {
            ripple: {
                colors: {
                    map: {
                        one: '#000',
                        two: '#fff',
                    },
                },
                darken: 0.2
            },
        },
    }).then(css => {
        expect(css).toMatchCss(`
            .ripple-bg-map-one {
                background-color: #000;
                background-position: center;
                transition: background 0.8s;
            }

            .ripple-bg-map-one:hover {
                background: #000000 radial-gradient(circle, transparent 1%, #000000 1%) center/15000%;
            }

            .ripple-bg-map-one:active {
                background-color: #000;
                background-size: 100%;
                transition: background 0s;
            }

            .ripple-bg-map-two {
                background-color: #fff;
                background-position: center;
                transition: background 0.8s;
            }

            .ripple-bg-map-two:hover {
                background: #CCCCCC radial-gradient(circle, transparent 1%, #CCCCCC 1%) center/15000%;
            }

            .ripple-bg-map-two:active {
                background-color: #fff;
                background-size: 100%;
                transition: background 0s;
            }
            `);
    });
});

test('The plugin rejects a dud darken value and uses the default', () => {
    return generatePluginCss({
        theme: {
            ripple: {
                colors: {
                    map: {
                        one: '#000',
                        two: '#fff',
                    },
                },
                darken: 'dudvalue'
            },
        },
    }).then(css => {
        expect(css).toMatchCss(`
            .ripple-bg-map-one {
                background-color: #000;
                background-position: center;
                transition: background 0.8s;
            }

            .ripple-bg-map-one:hover {
                background: #000000 radial-gradient(circle, transparent 1%, #000000 1%) center/15000%;
            }

            .ripple-bg-map-one:active {
                background-color: #000;
                background-size: 100%;
                transition: background 0s;
            }

            .ripple-bg-map-two {
                background-color: #fff;
                background-position: center;
                transition: background 0.8s;
            }

            .ripple-bg-map-two:hover {
                background: #CCCCCC radial-gradient(circle, transparent 1%, #CCCCCC 1%) center/15000%;
            }

            .ripple-bg-map-two:active {
                background-color: #fff;
                background-size: 100%;
                transition: background 0s;
            }
            `);
    });
});

test('The plugin rejects too low a darken value and uses the default', () => {
    return generatePluginCss({
        theme: {
            ripple: {
                colors: {
                    map: {
                        one: '#000',
                        two: '#fff',
                    },
                },
                darken: -0.5
            },
        },
    }).then(css => {
        expect(css).toMatchCss(`
            .ripple-bg-map-one {
                background-color: #000;
                background-position: center;
                transition: background 0.8s;
            }

            .ripple-bg-map-one:hover {
                background: #000000 radial-gradient(circle, transparent 1%, #000000 1%) center/15000%;
            }

            .ripple-bg-map-one:active {
                background-color: #000;
                background-size: 100%;
                transition: background 0s;
            }

            .ripple-bg-map-two {
                background-color: #fff;
                background-position: center;
                transition: background 0.8s;
            }

            .ripple-bg-map-two:hover {
                background: #CCCCCC radial-gradient(circle, transparent 1%, #CCCCCC 1%) center/15000%;
            }

            .ripple-bg-map-two:active {
                background-color: #fff;
                background-size: 100%;
                transition: background 0s;
            }
            `);
    });
});

test('The plugin rejects too high a darken value and uses the default', () => {
    return generatePluginCss({
        theme: {
            ripple: {
                colors: {
                    map: {
                        one: '#000',
                        two: '#fff',
                    },
                },
                darken: 1.2
            },
        },
    }).then(css => {
        expect(css).toMatchCss(`
            .ripple-bg-map-one {
                background-color: #000;
                background-position: center;
                transition: background 0.8s;
            }

            .ripple-bg-map-one:hover {
                background: #000000 radial-gradient(circle, transparent 1%, #000000 1%) center/15000%;
            }

            .ripple-bg-map-one:active {
                background-color: #000;
                background-size: 100%;
                transition: background 0s;
            }

            .ripple-bg-map-two {
                background-color: #fff;
                background-position: center;
                transition: background 0.8s;
            }

            .ripple-bg-map-two:hover {
                background: #CCCCCC radial-gradient(circle, transparent 1%, #CCCCCC 1%) center/15000%;
            }

            .ripple-bg-map-two:active {
                background-color: #fff;
                background-size: 100%;
                transition: background 0s;
            }
            `);
    });
});

test('The plugin will accept transition modifiers', () => {
    return generatePluginCss({
        theme: {
            ripple: {
                colors: {
                    map: {
                        one: '#000',
                        two: '#fff',
                    },
                },
                modifierTransition: 'background 0.2s',
                activeTransition: 'background 0.1s'
            },
        },
    }).then(css => {
        expect(css).toMatchCss(`
                .ripple-bg-map-one {
                    background-color: #000;
                    background-position: center;
                    transition: background 0.2s;
                }

                .ripple-bg-map-one:hover {
                    background: #000000 radial-gradient(circle, transparent 1%, #000000 1%) center/15000%;
                }

                .ripple-bg-map-one:active {
                    background-color: #000;
                    background-size: 100%;
                    transition: background 0.1s;
                }

                .ripple-bg-map-two {
                    background-color: #fff;
                    background-position: center;
                    transition: background 0.2s;
                }

                .ripple-bg-map-two:hover {
                    background: #CCCCCC radial-gradient(circle, transparent 1%, #CCCCCC 1%) center/15000%;
                }

                .ripple-bg-map-two:active {
                    background-color: #fff;
                    background-size: 100%;
                    transition: background 0.1s;
                }
            `);
    });
});

test('The plugin will use default transition modifiers', () => {
    return generatePluginCss({
        theme: {
            ripple: {
                colors: {
                    map: {
                        one: '#000',
                        two: '#fff',
                    },
                },
                modifierTransition: 'background 0.2s'
            },
        },
    }).then(css => {
        expect(css).toMatchCss(`
                .ripple-bg-map-one {
                    background-color: #000;
                    background-position: center;
                    transition: background 0.2s;
                }

                .ripple-bg-map-one:hover {
                    background: #000000 radial-gradient(circle, transparent 1%, #000000 1%) center/15000%;
                }

                .ripple-bg-map-one:active {
                    background-color: #000;
                    background-size: 100%;
                    transition: background 0s;
                }

                .ripple-bg-map-two {
                    background-color: #fff;
                    background-position: center;
                    transition: background 0.2s;
                }

                .ripple-bg-map-two:hover {
                    background: #CCCCCC radial-gradient(circle, transparent 1%, #CCCCCC 1%) center/15000%;
                }

                .ripple-bg-map-two:active {
                    background-color: #fff;
                    background-size: 100%;
                    transition: background 0s;
                }
            `);
    });
});


index.js
const _ = require('lodash');
const Color = require('color');

module.exports = function() {
    return ({ theme, e, addComponents, postcss }) => {
        const defaultBgColors = {};
        const rippleBgColors = theme('ripple.colors', defaultBgColors);
        const defaultDarkenValue = 0.2;
        let darkenValue = theme('ripple.darken', defaultDarkenValue);
        if (isNaN(darkenValue) || darkenValue > 1 || darkenValue < 0) {
            darkenValue = defaultDarkenValue;
        }
        const defaultModifierTransition = 'background 0.8s';
        let modifierTransition = theme('ripple.modifierTransition', defaultModifierTransition);
        const defaultActiveTransition = 'background 0s';
        let activeTransition = theme('ripple.activeTransition', defaultActiveTransition);

        function returnColorPair([modifier, value]) {
            return [
                `${modifier}`,
                {
                    backgroundColor: value,
                    backgroundPosition: 'center',
                    transition: modifierTransition,
                    '&:hover': {
                        background: `${Color(value)
                            .darken(darkenValue)
                            .hex()
                            .toString()} radial-gradient(circle, transparent 1%, ${Color(
                            value
                        )
                            .darken(darkenValue)
                            .hex()
                            .toString()} 1%) center/15000%`,
                    },
                    '&:active': {
                        backgroundColor: value,
                        backgroundSize: '100%',
                        transition: activeTransition,
                    },
                },
            ];
        }

        const allTheColors = _(rippleBgColors)
            .flatMap((value, modifier) => {
                if (typeof value == 'object') {
                    return _.map(value, (v, m) => {
                        return [`.${e(`ripple-bg-${modifier}-${m}`)}`, v];
                    });
                }
                if (
                    typeof value == 'string' &&
                    value.length > 1
                ) {
                    try {
                        Color(value)
                    } catch (err) {
                        return [];
                    }
                    return [[`.${e(`ripple-bg-${modifier}`)}`, value]];
                }
            })
            .value();

        const components = _.fromPairs(
            _.map(allTheColors, (color, index) => {
                return returnColorPair(color);
            })
        );

        addComponents(components);
    };
};
